modules/standard/CTypes.chpl

```diff
@@ -727,39 +725,158 @@ module CTypes {
     return c_pointer_return_const(arr[arr.domain.low]);
   }
 
+  /*
+    Returns a :type:`c_ptr` to the underlying buffer of a :type:`~String.string`
+
+    Note that the existence of this ``c_ptr`` has no impact on the lifetime of
+    the ``string``.  The returned pointer will be invalid if the ``string`` is
+    freed or even reallocated.
+
+    Halts if the ``string`` is empty and bounds checking is enabled.
+  */
+  inline proc c_ptrTo(ref s: string): c_ptr(c_uchar) {
+    if boundsChecking {
+      if (s.buffLen == 0) then
+        halt("Can't create a C pointer for an empty string.");
+    }
+    return c_pointer_return(s.buff[0]);
+  }
+
+  /*
+   Like ``c_ptrTo`` for :type:`~String.string`, but returns a :type:`c_ptrConst`
+   which disallows direct modification of the pointee.
+   */
+  inline proc c_ptrToConst(const ref s: string): c_ptrConst(c_uchar) {
+    if boundsChecking {
+      if (s.buffLen == 0) then
+        halt("Can't create a C pointer for an empty string.");
+    }
+    return c_pointer_return_const(s.buff[0]);
+  }
```

Comment:
@jeremiah-corrado : I hadn't realized these routines returned `c_ptrConst(c_uchar)` rather than `c_ptrConst(c_char)` until it came up in a side conversation today.  What's the rationale for that?  I.e., I would've thought `const char*` would be the obvious type for these rather than `const unsigned char*` (or `const signed char*`, for that matter).

Reply:
I believe I used `c_uchar` because `string ` and `bytes` use `uint(8)` for their buffers, so I wanted to keep things consistent. @arezaii and I discussed this briefly offline as well and determined that it could probably be switched to `c_char`. IIRC he mentioned switching the type to `c_ptrConst(c_char)` as a part of his `c_string` deprecation PR.

Reply:
🤔 

I see, that's a good point—I was clearly focused more on the destination of these calls (i.e., where I'd expect to pass them in C) than the source...  Huh, I'm not sure what to do here, especially given that our strings aren't really C-style ASCII strings anyway...

Maybe it's more important and appropriate that the "make me a deep-copy NULL-terminated C string from this string/bytes" method that we were discussing this week return ` c_ptrConst(c_char)` than it is that these routines do so?

Reply:
Yeah that sounds appropriate to me. I.e., `c_ptrToConst(myString)` could just mean "get me a pointer to this Chapel string's buffer", while `myString.getBuffer()` (or whatever the name is) would mean, get me a "C-compatible pointer to this string's buffer".